crates/cext/src/transpiler/transpile_function.rs

```diff
@@ -76,6 +78,124 @@ pub extern "C" fn qk_transpiler_default_options() -> TranspileOptions {
+///   for more details on the default values.
+/// @param layout A pointer to a pointer to a ``QkTranspileLayout`` object. On a successful
+///   execution (return code 0) a pointer to the layout object created transpiler will be written
+///   to this pointer.
+/// @param error A pointer to a pointer with an nul terminated string with an error description.
+///   If the transpiler fails a pointer to the string with the error description will be written
+///   to this pointer. That pointer needs to be freed with ``qk_str_free```. This can be a null
+///   pointer in which case the error will not be written out.
+///
+/// @returns The return code for the transpiler, ``QkExitCode_Success`` means success and all
+///   other values indicate an error.
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag``, ``target``, or ``layout``, are not valid, non-null
+/// pointers to a ``QkDag``, ``QkTarget``, or a ``QkTranspileLayout`` pointer
+/// respectively. ``options`` must be a valid pointer a to a ``QkTranspileOptions`` or ``NULL`.
+/// ``error`` must be a valid pointer to a ``char`` pointer or ``NULL``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpile_stage_init(
+    dag: *mut DAGCircuit,
+    target: *const Target,
+    options: *const TranspileOptions,
+    layout: *mut *mut TranspileLayout,
+    error: *mut *mut c_char,
+) -> ExitCode {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let dag = unsafe { mut_ptr_as_ref(dag) };
+    let target = unsafe { const_ptr_as_ref(target) };
+    let mut out_layout = TranspileLayout::new(
+        None,
+        None,
+        dag.qubits().objects().to_owned(),
+        dag.num_qubits() as u32,
+        dag.qregs().to_vec(),
+    );
+    let options = if options.is_null() {
+        &TranspileOptions::default()
+    } else {
+        // SAFETY: We checked the pointer is not null, then, per documentation, it is a valid
+        // and aligned pointer.
+        unsafe { const_ptr_as_ref(options) }
+    };
+
+    let approximation_degree = if options.approximation_degree.is_nan() {
+        None
+    } else {
+        if !(0.0..=1.0).contains(&options.approximation_degree) {
+            panic!(
+                "Invalid value provided for approximation degree, only NAN or values between 0.0 and 1.0 inclusive are valid"
+            );
+        }
+        Some(options.approximation_degree)
+    };
+    let mut commutation_checker = get_standard_commutation_checker();
+
+    match init_stage(
+        dag,
+        target,
+        options.optimization_level.into(),
+        approximation_degree,
+        &mut out_layout,
+        &mut commutation_checker,
+    ) {
+        Ok(_) => {
+            // SAFETY: Per the documentation result is a non-null aligned pointer to a pointer to
+            // a QKTranspileLayout
+            unsafe {
+                *layout = Box::into_raw(Box::new(out_layout));
```

Comment:
Should we use ``layout.write(out_layout)`` here instead?

Reply:
We can but it doesn't really change much here. But it also points out to me that I need to write a bit more detail in the safety docs about what happens to the existing pointer for `layout` because we don't free it so it needs to managed by the user if the inner pointer has data.